modules/widgets/migrate-to-core/gallery-widget.php

```diff
@@ -0,0 +1,192 @@
+			jetpack_migrate_gallery_widget_bump_stats( 'gallery-widget-skipped' );
+			continue;
+		}
+
+		$media_gallery[ $new_id ] = $widget_copy;
+
+		$sidebars_widgets = jetpack_migrate_gallery_widget_update_sidebars( $sidebars_widgets, $id, $new_id );
+
+		$widgets_to_unregister[ $id ] = $new_id;
+	}
+
+	if ( update_option( 'widget_media_gallery', $media_gallery ) ) {
+		delete_option( 'widget_gallery' );
+
+		// Now un-register old widgets and register new.
+		foreach ( $widgets_to_unregister as $id => $new_id ) {
+			wp_unregister_sidebar_widget( "gallery-${id}" );
+
+			// register new widget.
+			$media_gallery_widget = new WP_Widget_Media_Gallery();
+			$media_gallery_widget->_set( $new_id );
+			$media_gallery_widget->_register_one( $new_id );
+		}
+
+		wp_set_sidebars_widgets( $sidebars_widgets );
+
+		// Log if we migrated all, or some for this site.
+		foreach ( $widgets_to_unregister as $w ) {
+			jetpack_migrate_gallery_widget_bump_stats( 'gallery-widget-migrated' );
+		}
+
+		// We need to refresh on widgets page for changes to take effect.
+		// The jetpack_refresh_on_widget_page function is already defined in migrate-to-core/image-widget.php
+		add_action( 'current_screen', 'jetpack_refresh_on_widget_page' );
+	} else {
+		$widget_media_gallery = get_option( 'widget_media_gallery' );
+		if ( is_array( $widget_media_gallery ) ) {
+			delete_option( 'widget_gallery' );
+		}
+	}
+	Jetpack_Options::update_option( 'gallery_widget_migration', true );
+}
+
+function jetpack_migrate_gallery_widget_is_importable( $widget ) {
+	// Can be caused by instantiating but not populating a widget in the Customizer.
+	if ( empty( $widget ) ) {
+		return false;
+	}
+	// The array as stored in the option constains two keys and one
+	// is a string `_multiwidget` which does not represent a widget, so we skip it
+	if ( ! is_array( $widget ) ) {
+		return false;
+	}
+	return true;
+}
+
+/**
+ * Returns a transformed version of the Gallery Widget.
+ * Will return null if the widget is either empty, is not an array or has more keys than expected
+ *
+ * @param $widget   One of the Jetpack Gallery widgets to be transformed into a new Core Media Gallery Widget
+ *
+ * @return array|null
+ */
+function jetpack_migrate_gallery_widget_upgrade_widget( $widget ) {
+	$whitelisted_keys = array(
+		'ids' => '',
+		'link' => '',
+		'title' => '',
+		'type' => '',
+		'random' => '',
+		'conditions' => '',
+	);
+
+	$default_data = array(
+		'columns' => 3,
+		'ids' => array(),
+		'link_type' => '',
+		'orderby_random' => false,
+		'size' => 'thumbnail',
+		'title' => '',
+		'type' => '',
+	);
+
+	if ( ! jetpack_migrate_gallery_widget_is_importable( $widget ) ) {
+		return null;
+	}
+	// Ensure widget has no keys other than those expected.
+	// Not all widgets have conditions, so lets add it in.
+	$widget_copy = array_merge( array( 'conditions' => null ), $widget );
+	if ( count( array_diff_key( $widget_copy, $whitelisted_keys ) ) > 0 ) {
+		return null;
```

Comment:
I would add some sort of alert/logging here.  If we're seeing unexpected values we should evaluate and understand them.